File: TddExampleCpp/IndexerTest/IndexerTest.cpp
#include "pch.h"
#include "CppUnitTest.h"

#include "../Indexer/IndexerCpp.h"
#include "../Indexer/IndexerCpp.cpp"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace IndexerTest
{
	TEST_CLASS(IndexerTest)
	{
	private:
		double* array = new double[5]{ 1,2,3,4,5 };

	public:
		
		TEST_METHOD(HaveCorrectLength)
		{
			IndexerCpp indexer(array, 5, 1, 2);
			Assert::AreEqual(2, indexer.getLength());
		}

		TEST_METHOD(GetCorrectly)
		{
			IndexerCpp indexer(array, 4, 1, 2);
			Assert::AreEqual(2.0, indexer[0]);
			Assert::AreEqual(3.0, indexer[1]);
		}

		TEST_METHOD(SetCorrectly)
		{
			IndexerCpp indexer(array, 4, 1, 2);
			indexer[0] = 10;
			Assert::AreEqual(10.0, array[1]);
		}

		TEST_METHOD(IndexerDoesNotCopyArray)
		{
			IndexerCpp indexer1(array, 4, 1, 2);
			IndexerCpp indexer2(array, 4, 0, 2);
			indexer1[0] = 12345;
			Assert::AreEqual(12345.0, indexer2[1]);
		}

		TEST_METHOD(FailWithWrongArguments1)
		{
			auto func = []() {
				double* arr = new double[4]{ 1, 2, 3, 4 };
				IndexerCpp temp(arr, 4, -1, 3);
			};
			Assert::ExpectException<std::invalid_argument>(func);
		}
		TEST_METHOD(FailWithWrongArguments2)
		{
			auto func = []() {
				double* arr = new double[4]{ 1, 2, 3, 4 };
				IndexerCpp temp(arr, 4, 1, -1);
			};
			Assert::ExpectException<std::invalid_argument>(func);
		}
		TEST_METHOD(FailWithWrongArguments3)
		{
			auto func = []() {
				double* arr = new double[4]{ 1, 2, 3, 4 };
				IndexerCpp temp(arr, 4, 1, 10);
			};
			Assert::ExpectException<std::invalid_argument>(func);
		}
		TEST_METHOD(FailWithWrongIndexing1)
		{
			auto func = []() {
				double* arr = new double[4]{ 1, 2, 3, 4 };
				IndexerCpp indexer(arr, 4, 1, 2);
				double temp = indexer[-1];
			};
			Assert::ExpectException<std::out_of_range>(func);
		}
		TEST_METHOD(FailWithWrongIndexing2)
		{
			auto func = []() {
				double* arr = new double[4]{ 1, 2, 3, 4 };
				IndexerCpp indexer(arr, 4, 1, 2);
				double tmp = indexer[10];
			};
			Assert::ExpectException<std::out_of_range>(func);
		}



	};
}

File: TddExample/IndexerTest/IndexerTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using Indexer1;

namespace IndexerTest
{
    [TestClass]
    public class IndexerTest
    {
        double[] array = new double[] { 1, 2, 3, 4 };


        [TestMethod]
        public void HaveCorrectLength()
        {
            var indexer = new Indexer(array, 1, 2);
            Assert.AreEqual(2, indexer.Length);
        }

        [TestMethod]
        public void GetCorrectly()
        {
            var indexer = new Indexer(array, 1, 2);
            Assert.AreEqual(2, indexer[0]);
            Assert.AreEqual(3, indexer[1]);
        }

        [TestMethod]
        public void SetCorrectly()
        {
            var indexer = new Indexer(array, 1, 2);
            indexer[0] = 10;
            Assert.AreEqual(10, array[1]);
        }

        [TestMethod]
        public void IndexerDoesNotCopyArray()
        {
            var indexer1 = new Indexer(array, 1, 2);
            var indexer2 = new Indexer(array, 0, 2);
            indexer1[0] = 100500;
            Assert.AreEqual(100500, indexer2[1]);
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentException))]
        public void FailWithWrongArguments1()
        {
            Assert.Equals(typeof(ArgumentException), new Indexer(array, -1, 3));
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentException))]
        public void FailWithWrongArguments2()
        {
            Assert.Equals(typeof(ArgumentException), new Indexer(array, 1, -1));
        }

        [TestMethod]
        [ExpectedException(typeof(ArgumentException))]
        public void FailWithWrongArguments3()
        {
            Assert.Equals(typeof(ArgumentException), new Indexer(array, 1, 10));
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void FailWithWrongIndexing1()
        {
            var indexer = new Indexer(array, 1, 2);
            Assert.Equals(typeof(IndexOutOfRangeException), indexer[-1]);
        }

        [TestMethod]
        [ExpectedException(typeof(IndexOutOfRangeException))]
        public void FailWithWrongIndexing2()
        {
            var indexer = new Indexer(array, 1, 2);
            Assert.AreNotEqual(typeof(IndexOutOfRangeException), indexer[10]);
        }
    }
}
